pulp_2to3_migration/app/pre_migration.py

```diff
@@ -623,12 +632,17 @@ def delete_old_resources(plan):
     with transaction.atomic():
         pulp2distributors_with_old_distributions_qs = Pulp2Distributor.objects.filter(
             old_dist_query).only('pulp3_distribution')
-
         for pulp2distributor in pulp2distributors_with_old_distributions_qs:
             if pulp2distributor.is_migrated:
                 pulp2distributor.is_migrated = False
                 pulp2distributor.save()
             if pulp2distributor.pulp3_publication:
+                # todo check when publication is shared by multiple distro
+                pulp2d=pulp2distributor.pulp3_publication.pulp2distributor_set.all()
+                for d in pulp2d:
+                    d.is_migrated=False
+                    d.save()
+                    d.pulp3_distribution.delete()
                 pulp2distributor.pulp3_publication.delete()
-            if pulp2distributor.pulp3_distribution:
+            elif hasattr(pulp2distributor, 'pulp3_distribution') and pulp2distributor.pulp3_distribution:
```

Comment:
How is it possible not to have this attribute?
It can be null/None but it should be present I think https://github.com/pulp/pulp-2to3-migration/blob/master/pulp_2to3_migration/app/models/repository.py#L158

Reply:
i get a keyerror when i iterate trough the qs.
```
pulp2distributor.pulp3_distribution
Traceback (most recent call last):
  File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/fields/related_descriptors.py", line 164, in __get__
    rel_obj = self.field.get_cached_value(instance)
  File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/fields/mixins.py", line 13, in get_cached_value
    return instance._state.fields_cache[cache_name]
KeyError: 'pulp3_distribution'
During handling of the above exception, another exception occurred:
``` 
It happens because i have already removed the distribution from that pulp2distributor on line #L646

Reply:
Something is off here. Maybe because you remove them inside the loop.
@dralley , any thoughts?

Reply:
We shouldn't be using ```.only()``` if we're reading from other fields as well.  I think that might have snuck through various refactors.  I'm not sure why it outright fails like this though, it should just be doing extra queries behind the scenes, so this failure actually kinda seems like a Django bug.

I would definitely try removing the ```.only()``` and see if that fixes it though.  It serves no useful purpose here/anymore.

Reply:
so removing `only()` did not help, and some weird stuff is ongoing.

I have removed all my changes from lines 640-647 and had such plan
```
{ "plan": {
   "plugins":[
      {
         "type":"iso",
         "repositories":[
            {
               "name":"file",
               "repository_versions":[
                  {
                     "pulp2_repository_id":"file",
                     "pulp2_distributor_repository_ids":[
                        "file"
                     ]
                  }
               ],
               "pulp2_importer_repository_id":"file"
            },
            {
               "name":"new-name",
               "repository_versions":[
                  {
                     "pulp2_repository_id":"file2",
                     "pulp2_distributor_repository_ids":[
                        "file2",
                        "file-many"
                     ]
                  }
               ]
            }
         ]
      }
   ]
}}
```

I ran the migration, then i removed and re-created file2 and file-many repos. and Re-run the migrations

As a result i got a key error on line L639
```
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: pulpcore.app.models.publication.Publication.DoesNotExist: Publication matching query does not exist.
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: self.model._meta.object_name
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/query.py", line 408, in get
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: return qs.get(self.field.get_reverse_related_filter(instance))
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/fields/related_descriptors.py", line 145, in get_object
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: rel_obj = self.get_object(instance)
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/fields/related_descriptors.py", line 178, in __get__
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: pulp2distributor.pulp3_publication
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "<ipython-input-4-80287837a1a2>", line 1, in <module>
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: exec(code_obj, self.user_global_ns, self.user_ns)
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "/usr/local/lib/pulp/lib64/python3.6/site-packages/IPython/core/interactiveshell.py", line 3331, in run_code
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: Traceback (most recent call last):
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: During handling of the above exception, another exception occurred:
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: KeyError: 'pulp3_publication'
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: return instance._state.fields_cache[cache_name]
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/fields/mixins.py", line 13, in get_cached_value
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: rel_obj = self.field.get_cached_value(instance)
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: File "/usr/local/lib/pulp/lib64/python3.6/site-packages/django/db/models/fields/related_descriptors.py", line 164, in __get__
Jun 09 18:36:35 pulp2-nightly-pulp3-source-centos7.fluffy.example.com rq[24082]: Traceback (most recent call last):
```

Reply:
@ipanova It seems like what is happening is:

* ```pulp3distribution``` gets deleted
* ```pulp2distributor.pulp3distribution``` is accessed.
    * Internally, inside Django, it misses the cache, which is *normal*.  At this point it's still just a PK since Django doesn't eagerly pre-load data across table relationships unless you use ```prefetch_related```
    * Django performs a query using the cached PK, which throws ```DoesNotExist```

Reply:
Looks like you already found the solution I was about to suggest :D